examples/src/Auction.hs

```diff
@@ -0,0 +1,303 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# OPTIONS_GHC -fno-ignore-interface-pragmas #-}
+{-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
+{-# OPTIONS_GHC -fno-specialise #-}
+
+-- These language extensions are just what Split.hs uses, I honestly
+-- have no idea what most of them do or if I actually need them.
+
+-- | Arrange an auction with a preset deadline and minimum bid.
+module Auction where
+
+import qualified Ledger as L
+import qualified Ledger.Ada as Ada
+import qualified Ledger.Interval as Interval
+import Ledger.Typed.Scripts as Scripts
+import qualified Ledger.Value as Value
+import qualified PlutusTx
+import PlutusTx.Prelude
+import qualified Prelude as Haskell
+
+-- * Data types
+
+-- | All the statically known data associated with an auction
+data Parameters' = Parameters'
+  { -- | no bids are accepted later than this
+    bidDeadline' :: L.POSIXTime,
+    -- | minimum bid in Lovelace
+    minBid' :: Integer,
+    -- | value that is being auctioned
+    lot' :: L.Value
+  }
+  deriving (Haskell.Show)
+
+-- | All data associated with an auction, including the data only
+-- known after the opening transaction
+data Parameters = Parameters
+  { staticParameters :: Parameters',
+    -- | address of the seller
+    seller :: L.PubKeyHash,
+    -- | autref of the utxo the seller spent the lot from (needed as a parameter for the minting policy of the thread token)
+    lotOutRef :: L.TxOutRef,
```

Comment:
Due to the tutorial nature of this example and because you only use this `TxOutRef` in the off-chain code, I think we should remove it from the script parameters: it might get confusing to future readers why that is, especially because we're already passing the `threadTokenAssetClass` below.

On a similar note, we should add a comment to the `threadTokenAssetClass` explaining why is it there and why it has to be that way: to break the mutual dependency cycle between the minting policy and the validator